keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -183,70 +184,78 @@ class ApplicationService(
 
     val environmentSummaries = getEnvironmentSummariesFor(application)
 
-    return deliveryConfig.artifacts.map { artifact ->
+    val artifactSummaries = deliveryConfig.artifacts.map { artifact ->
       val artifactVersionSummaries = repository.artifactVersions(artifact).map { version ->
         val artifactSummariesInEnvironments = mutableSetOf<ArtifactSummaryInEnvironment>()
 
         environmentSummaries.forEach { environmentSummary ->
           val environment = deliveryConfig.environments.find { it.name == environmentSummary.name }!!
-          environmentSummary.getArtifactPromotionStatus(artifact, version)?.let { status ->
-            var artEnvSummary = when (status) {
-              PENDING -> ArtifactSummaryInEnvironment(
-                environment = environmentSummary.name,
-                version = version,
-                state = status.name.toLowerCase()
-              )
-              SKIPPED -> {
-                // some environments contain relevant info for skipped artifacts, so
-                // try and find that summary before defaulting to less information
-                val potentialSummary = repository.getArtifactSummaryInEnvironment(
-                  deliveryConfig = deliveryConfig,
-                  environmentName = environmentSummary.name,
-                  artifactReference = artifact.reference,
-                  version = version
-                )
-                if (potentialSummary == null || potentialSummary.state == "pending") {
-                  ArtifactSummaryInEnvironment(
-                    environment = environmentSummary.name,
-                    version = version,
-                    state = status.name.toLowerCase()
+          environmentSummary.getArtifactPromotionStatus(artifact, version)
```

Comment:
The logic in this file is unchanged. I just extracted some of it to a new function to make the loops easier to reason about.